File: internal/handlers/comments.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strconv"

	"real-time-forum/internal/database"
	"real-time-forum/internal/models"
)

type CreateCommentRequest struct {
	PostID  int    `json:"postId"`
	Content string `json:"content"`
}

func Comments(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodGet {
		postID, err := strconv.Atoi(r.URL.Query().Get("post_id"))
		if err != nil || postID <= 0 {
			SendError(w, http.StatusBadRequest, "Invalid post ID!")
			return
		}

		comments, err := database.GetCommentsByPost(postID)
		if err != nil {
			SendError(w, http.StatusInternalServerError, "Failed to get comments!")
			return
		}

		SendResponse(w, http.StatusOK, comments)
		return
	}

	if r.Method == http.MethodPost {
		userID, ok := r.Context().Value("userID").(int)
		if !ok {
			SendError(w, http.StatusUnauthorized, "Unauthorized!")
			return
		}

		var req CreateCommentRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			SendError(w, http.StatusBadRequest, "Invalid JSON Data!")
			return
		}

		if req.PostID <= 0 || req.Content == "" {
			SendError(w, http.StatusBadRequest, "Post ID and Content are required!")
			return
		}

		comment := models.Comment{
			UserID:  userID,
			PostID:  req.PostID,
			Content: req.Content,
		}

		id, err := database.CreateComment(comment)
		if err != nil {
			SendError(w, http.StatusInternalServerError, "Failed to create comment!")
			return
		}

		SendResponse(w, http.StatusCreated, map[string]any{
			"success": true,
			"id":      id,
		})
		return
	}

	SendError(w, http.StatusMethodNotAllowed, "Method Not Allowed!")
}

File: web/js/api/comments.js
export async function getComments(postID) {
    try {
        const res = await fetch(`/api/comments?post_id=${postID}`)
        if (!res.ok) return []
        return await res.json()
    } catch (error) {
        console.log(error)
        return []
    }
}

export async function createComment(postId, content) {
    try {
        const res = await fetch("/api/comments", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                postId: parseInt(postId, 10),
                content: content
            })
        });
        const data = await res.json();
        if (!res.ok) {
            return { error: data.error || "Failed to create comment" };
        }
        return data;
    } catch (error) {
        return { error: "Network connection error. Please try again." };
    }
}
